### PR TITLE
feat(actions): add expression/calculation support to Telemetry Display templates (#192)

### LINKED ISSUE #192: feat(actions): add expression/calculation support to Telemetry Display templates
### Problem

The Telemetry Display action currently supports Mustache-style variable substitution (`{{telemetry.Speed}}`, `{{self.position}}`), but users cannot perform calculations or transformations on the values. This limits the action to displaying raw telemetry values only.

**Example use case:** A user wants to display fuel to add in gallons, but `dpFuelAddKg` is in kilograms and `DriverCarFuelKgPerLtr` is the conversion factor. Today there's no way to compute:

```text
{{telemetry.dpFuelAddKg}} / ({{sessionInfo.DriverInfo.DriverCarFuelKgPerLtr}} * 0.264172)
```

### Proposed Solution

Add expression evaluation support to Telemetry Display templates, similar to [SimHub's NCalc integration](https://github.com/SHWotever/SimHub/wiki/NCalc-scripting---Introduction). This would allow users to write mathematical expressions that reference telemetry and session variables.

### Syntax

Use `{{= expr }}` delimiters — a natural extension of the existing `{{variable}}` Mustache-like syntax, with an `=` prefix to indicate an expression:

```text
{{= telemetry.dpFuelAddKg / (sessionInfo.DriverInfo.DriverCarFuelKgPerLtr * 0.264172) }}
```

### Minimum Viable Feature

- Arithmetic operators: `+`, `-`, `*`, `/`, `%`
- Parentheses for grouping
- Numeric literals
- Variable references (same dot-notation keys as existing templates)
- Result formatting: round to N decimal places (e.g., `round(expr, 2)` or a format specifier)

### Nice to Have

- Comparison operators for conditional display (`>`, `<`, `>=`, `<=`, `==`, `!=`)
- Ternary/conditional: `condition ? "A" : "B"`
- Built-in functions: `round()`, `floor()`, `ceil()`, `abs()`, `min()`, `max()`
- String concatenation
- Unit conversion helpers (kg→lbs, km/h→mph, L→gal, C→F)

### Affected Areas

- [ ] **Template resolver** (`packages/iracing-sdk/src/template-resolver.ts`) — add expression parsing/evaluation alongside existing variable substitution
- [ ] **Template resolver tests** — comprehensive tests for expression evaluation, edge cases, error handling
- [ ] **Telemetry Display action** (`packages/actions/src/actions/telemetry-display.ts`) — no changes expected if resolver handles it transparently
- [ ] **Documentation** (`packages/website/src/content/docs/`) — document expression syntax, available operators/functions, examples
- [ ] **Template variables docs** — add section on expressions with examples
- [ ] **PI template** — consider adding expression syntax hints/examples to the template textarea help text

### Security Considerations

The expression evaluator must **not** use `eval()` or `new Function()`. Use a safe expression parser that only supports the defined operators and functions — no arbitrary code execution.

### PR DESCRIPTION
**Pull request by niklam**
## Related Issue

Fixes #192

## What changed?

Templates in the Telemetry Display, Chat, and Race Admin actions now support safe expression evaluation with the `{{= expression }}` syntax alongside plain `{{variable}}` placeholders.

- New hand-rolled expression evaluator in `@iracedeck/iracing-sdk` (`expression-evaluator.ts`): tokenizer + recursive-descent parser + evaluator with a 200-entry AST cache. No `eval()`/`new Function()`, zero new dependencies, hardened against prototype-chain lookups and oversized input (1000-char cap).
- Full feature set: arithmetic `+ - * / %`, parentheses, numeric/string literals, dot-notation variable references, comparisons, ternary, `round/floor/ceil/abs/min/max`, and string concatenation via `+`.
- `TemplateContext` became a combined `{ display, raw }` object built in a single walk, so expressions compute on raw full-precision values while plain `{{variable}}` display output stays byte-identical. Zero changes to any action source file — Chat and Race Admin gain expressions transparently through the shared resolver.
- Hybrid error model: a syntax error leaves the raw `{{= … }}` text visible on the key (typo feedback); evaluation problems (unknown variable, division by zero) render as empty text.
- Result formatting matches existing display conventions: integers bare, floats 2 decimals by default, `round(x, n)` renders exactly n decimals, booleans render Yes/No.
- Docs: new Expressions section on the website template-variables page, Telemetry Display action page update, PI help blurbs (Telemetry Display, Chat, Race Admin), skill updates, and a design doc under `docs/plans/`. The issue's fuel example formula was corrected during implementation (kg ÷ (kg/L × 3.78541) — the issue's original constant overstated gallons ~14×).

## How to test

1. Add a Telemetry Display key with template `{{= round(telemetry.dpFuelAddKg / (sessionInfo.DriverInfo.DriverCarFuelKgPerLtr * 3.78541), 1) }}` and verify it shows fuel-to-add in gallons, updating live.
2. Try `{{= round(telemetry.Speed * 3.6, 0) }}` for km/h and `{{= telemetry.OnPitRoad == 1 ? 'PIT' : '' }}` for conditional display.
3. Type a deliberate syntax error (e.g. `{{= round(telemetry.Speed, }}`) and verify the raw text stays visible on the key.
4. Use a `{{= … }}` expression inside a Chat message template and verify it resolves when sent.
5. `pnpm build` (18/18) and `pnpm test` (5268 tests) pass.

## Checklist

- [x] Linked to an approved issue
- [x] New code has unit tests
- [x] Changes registered in all applicable plugins (Stream Deck, Mirabox)
- [x] No unrelated changes included

## Summary by CodeRabbit

* **New Features**
  * Added template expressions (`{{= ... }}`) for dynamic calculations in Chat, Race Admin, and Telemetry Display. Supports arithmetic, comparisons, ternary, and math functions (`round`, `floor`, `ceil`, `abs`, `min`, `max`).

* **Behavior Changes**
  * Expressions evaluate against raw full‑precision telemetry (not display-formatted values). Parse errors leave the placeholder verbatim; runtime errors render as an empty string.

* **Documentation**
  * UI help text and docs updated with syntax, examples, and limits.

### CODE REVIEW

File: packages/iracing-sdk/src/expression-evaluator.ts [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+  vars: Record<string, ExpressionValue>,
+): EvalResult {
+  const left = evaluateAst(node.left, vars).value;
+  const right = evaluateAst(node.right, vars).value;
+
+  switch (node.op) {
+    case "+":
+      if (typeof left === "string" || typeof right === "string") {
+        return { value: stringifyValue(left) + stringifyValue(right) };
+      }
```
[the remark this answers is not in the document]
Comment (niklam) [reply packages/iracing-sdk/src/expression-evaluator.ts]: Accepted — good catch, this broke the most natural unit-suffix pattern. `evaluateBinary` now keeps each operand's evaluation envelope and formats string-concatenation operands via `formatResult`, so `round(x, n)` keeps its requested precision inside concatenation (`round(51.24, 1) + ' km/h'` → `51.2 km/h`). Behavior for non-round operands is unchanged. Docs updated to match (template-variables.md formatting section + example, design doc). Fixed in 1b893205.

File: packages/iracing-sdk/src/expression-evaluator.test.ts [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+
+  it("should chain concatenations left-to-right", () => {
+    expect(resolveExpression("'Lap ' + 5 + '/' + 10", {})).toBe("Lap 5/10");
+  });
+
+  it("should stringify booleans as Yes/No", () => {
+    expect(resolveExpression("'DRS: ' + on", { on: true })).toBe("DRS: Yes");
+    expect(resolveExpression("'DRS: ' + on", { on: false })).toBe("DRS: No");
+  });
+});
```
[the remark this answers is not in the document]
Comment (niklam) [reply packages/iracing-sdk/src/expression-evaluator.test.ts:497]: Added in 1b893205 alongside the behavior fix: `round(1/3, 4) + 's'` → `0.3333s`, `'v=' + round(51.24, 1)` → `v=51.2`, plus the FuelLevel unit-suffix case (`42.4 L`).

File: packages/iracing-sdk/src/index.ts [hunk 3]
```diff
@@ -100,17 +100,17 @@ export {
...
   buildTemplateContextFromData,
-  flattenForDisplay,
   prefixKeys,
   splitDriverName,
   findNearestDriverOnTrack,
   findDriverByRacePosition,
   formatTimeRemaining,
   type TemplateContext,
+  type TemplateValue,
 } from "./template-context.js";
```
[the remark this answers is not in the document]
Comment (niklam) [reply packages/iracing-sdk/src/index.ts:114]: Declining this one. `@iracedeck/iracing-sdk` is a private workspace package consumed only inside this monorepo — it isn't published to npm, has no external consumers, and the project keeps no changelog/semver contract (release notes are generated from PR titles). Both removed symbols were `@internal Exported for testing` helpers whose last production caller disappeared in this very PR, and the only remaining textual references are historical design documents. Re-introducing deprecated re-export shims with runtime warnings would add dead code with zero consumer benefit, against the project's convention of removing orphaned exports outright (this PR's refactor commit did exactly that on reviewer guidance). If the SDK is ever published externally, a deprecation policy can be introduced then.

**Comment by niklam** [comment]
Both review-body nitpicks are addressed in 11e0e146: `MAX_EXPRESSION_LENGTH` is now exported from the evaluator and `TEMPLATE_PATTERN` is built from it (single source of truth for the cap), and resolver-level boundary tests now lock the documented contracts — over-limit expressions stay verbatim, and an expression terminates at the first `}}` even inside a string literal. Additionally, the Major concat-precision finding led to a behavior improvement in 1b893205: `round(x, n)` now keeps its precision when concatenated with strings, with docs updated to match.
